code/trans.c

```diff
@@ -0,0 +1,377 @@
+  return TRUE;
+}
+
+
+/* Allocator functions for the Table oldToNew */
+
+static void *transformTableAlloc(void *closure, size_t size)
+{
+  Transform transform = (Transform)closure;
+  Res res;
+  void *p;
+
+  AVERT(Transform, transform);
+
+  res = ControlAlloc(&p, transform->arena, size);
+  if (res != ResOK)
+    return NULL;
+
+  return p;
+}
+
+static void transformTableFree(void *closure, void *p, size_t size)
+{
+  Transform transform = (Transform)closure;
+  AVERT(Transform, transform);
+  ControlFree(transform->arena, p, size);
+}
+
+
+Res TransformCreate(Transform *transformReturn, Arena arena)
+{
+  Transform transform;
+  Res res;
+  void *p;
+
+  AVER(transformReturn != NULL);
+  AVERT(Arena, arena);
+
+  res = ControlAlloc(&p, arena, sizeof(TransformStruct));
+  if (res != ResOK)
+    goto failAlloc;
+  transform = (Transform)p;
+  
+  transform->oldToNew = NULL;
+  transform->arena = arena;
+  transform->epoch = ArenaEpoch(arena);
+  transform->aborted = FALSE;
+
+  transform->sig = TransformSig;
+
+  AVERT(Transform, transform);
+
+  res = TableCreate(&transform->oldToNew,
+                    0, /* don't grow table until TransformAddOldNew */
+                    transformTableAlloc,
+                    transformTableFree,
+                    transform,
+                    0, 1); /* these can't be old references */
+  if (res != ResOK)
+    goto failTable;
+
+  *transformReturn = transform;
+  return ResOK;
+
+failTable:
+  ControlFree(arena, transform, sizeof(TransformStruct));
+failAlloc:
+  return res;
+}
+
+
+void TransformDestroy(Transform transform)
+{
+  Arena arena;
+  Table oldToNew;
+
+  AVERT(Transform, transform);
+
+  /* TODO: Log some transform statistics. */
+
+  /* Workaround bootstrap problem, see .check.boot */
+  oldToNew = transform->oldToNew;
+  transform->oldToNew = NULL;
+  TableDestroy(oldToNew);
+
+  arena = TransformArena(transform);
+  transform->sig = SigInvalid;
+  ControlFree(arena, transform, sizeof(TransformStruct));
+}
+
+
+/* TransformArena -- return transform's arena
+ * 
+ * Must be thread-safe as it is called outside the arena lock. See
+ * <design/thread-safety/#sol.check>
+ */
+
+Arena TransformArena(Transform transform)
+{
+  Arena arena;
+  AVER(TESTT(Transform, transform));
+  arena = transform->arena;
+  AVER(TESTT(Arena, arena));
+  return arena;
+}
+
+
+Res TransformAddOldNew(Transform transform,
+                       Ref old_list[],
+                       Ref new_list[],
+                       Count count)
+{
+  Res res;
+  Index i;
+  Count added = 0;
+
+  AVERT(Transform, transform);
+  AVER(old_list != NULL);
+  AVER(new_list != NULL);
+  /* count: cannot check */
+  
+  res = TableGrow(transform->oldToNew, count);
+  if (res != ResOK)
+    return res;
+
+  for(i = 0; i < count; ++i) {
+    /* NOTE: If the mutator isn't adding references while the arena is parked,
+       we might need to access the client-provided lists, using ArenaRead. */
```

Comment:
M. If we're assuming it's parked, we should assert that it is and document the restriction. This isn't good enough for public code. rule.generic.achieve

Reply:
Fixed: Added assertions to ensure this assumption is true in 666345955